pkg/operators/common/resourcecache.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2024 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	oldObjectTTL = 2 * time.Second
+)
+
+type oldResource[T any] struct {
+	deletionTimestamp time.Time
+	obj               T
+}
+
+type ResourceCache[T any, Key comparable] struct {
+	sync.Mutex
+	current map[Key]*T
+	old     map[Key]oldResource[*T]
+	toKey   func(*T) (Key, error)
+}
+
+func NewResourceCache[T any, Key comparable](toKey func(*T) (Key, error)) *ResourceCache[T, Key] {
+	return &ResourceCache[T, Key]{
+		current: make(map[Key]*T),
+		old:     make(map[Key]oldResource[*T]),
+		toKey:   toKey,
+	}
+}
+
+func (c *ResourceCache[T, Key]) Clear() {
+	c.Lock()
+	defer c.Unlock()
+	c.current = make(map[Key]*T)
+	c.old = make(map[Key]oldResource[*T])
+}
+
+func (c *ResourceCache[T, Key]) Add(obj *T) {
+	key, err := c.toKey(obj)
+	if err != nil {
+		log.Warnf("ResourceCache: add resource: %v", err)
+		return
+	}
+	c.Lock()
+	defer c.Unlock()
+	c.current[key] = obj
+	delete(c.old, key)
+}
+
+func (c *ResourceCache[T, Key]) Remove(obj *T) {
+	key, err := c.toKey(obj)
+	if err != nil {
+		log.Warnf("ResourceCache: remove resource: %v", err)
+		return
+	}
+	c.Lock()
+	defer c.Unlock()
+	delete(c.current, key)
+	c.old[key] = oldResource[*T]{deletionTimestamp: time.Now(), obj: obj}
+}
+
+func (c *ResourceCache[T, Key]) PruneOldObjects() {
+	c.Lock()
+	defer c.Unlock()
+	now := time.Now()
+	for key, oldObj := range c.old {
+		if now.Sub(oldObj.deletionTimestamp) > oldObjectTTL {
+			delete(c.old, key)
+		}
+	}
+}
+
+func (c *ResourceCache[T, Key]) ToSlice() []*T {
+	c.PruneOldObjects()
+	c.Lock()
+	defer c.Unlock()
+
+	objs := make([]*T, 0, len(c.current)+len(c.old))
+	for _, obj := range c.current {
+		objs = append(objs, obj)
+	}
+	for key, oldObj := range c.old {
+		if _, ok := c.current[key]; !ok {
+			objs = append(objs, oldObj.obj)
+		}
+	}
+	return objs
+}
+
+func (c *ResourceCache[T, Key]) Get(key Key) *T {
+	c.PruneOldObjects()
```

Comment:
Could also be ran async on a timer of course. Will do that.

(Shouldn't be too heavy though, since old Resources are quite rare in our use cases maybe)